lib/method_profiler.rb

```diff
@@ -38,6 +38,56 @@ def #{method_name}(*args, &blk)
     klass.class_eval patches
   end
 
+  def self.patch_with_debug_sql(klass, methods, name, no_recurse: false)
+    patches = methods.map do |method_name|
+
+      recurse_protection = ""
+      if no_recurse
+        recurse_protection = <<~RUBY
+          return #{method_name}__mp_unpatched_debug_sql(*args, &blk) if @mp_recurse_protect_#{method_name}
+          @mp_recurse_protect_#{method_name} = true
+        RUBY
+      end
+
+      <<~RUBY
+      unless defined?(#{method_name}__mp_unpatched_debug_sql)
+        alias_method :#{method_name}__mp_unpatched_debug_sql, :#{method_name}
+        def #{method_name}(*args, &blk)
+          unless prof = Thread.current[:_method_profiler]
+            return #{method_name}__mp_unpatched_debug_sql(*args, &blk)
+          end
+          #{recurse_protection}
+
+          query = args[0]
+          should_filter = #{@@instrumentation_debug_sql_filter_transactions} &&
+                            (query == "COMMIT" || query == "BEGIN" || query == "ROLLBACK")
+          if !should_filter
+            puts "debugsql (sql): " + query
+          end
+
+          begin
+            start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+            #{method_name}__mp_unpatched_debug_sql(*args, &blk)
+          ensure
+            data = (prof[:#{name}] ||= {duration: 0.0, calls: 0})
+            duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start
+
+            if !should_filter
+              puts "debugsql (sec): " + duration.to_s
```

Comment:
a bit too much time fidelity here ....  `duration.round(3).to_s` should be plenty

Reply:
also ... we got to output to STDERR not STDOUT